publication-rest/src/main/java/no/unit/nva/publication/update/UpdatePublicationHandler.java

```diff
@@ -304,12 +307,45 @@ private Publication updateMetadata(UpdatePublicationRequest input,
         var customerApiClient = getCustomerApiClient();
         var customer = fetchCustomerOrFailWithBadGateway(customerApiClient, publicationUpdate.getPublisher().getId());
         validatePublication(publicationUpdate, customer);
+        injectFileUploadDetails(existingPublication, publicationUpdate, userInstance);
         setRrsOnFiles(publicationUpdate, existingPublication, customer, userInstance.getUsername(), permissionStrategy);
         upsertPublishingRequestIfNeeded(existingPublication, publicationUpdate, customer, requestInfo);
 
         return resourceService.updatePublication(publicationUpdate);
     }
 
+    private static void injectFileUploadDetails(Publication existingPublication, Publication publicationUpdate,
+                                                UserInstance userInstance) {
+        if (!existingPublication.getAssociatedArtifacts().equals(publicationUpdate.getAssociatedArtifacts())) {
+            var username = new Username(userInstance.getUsername());
+            var uploadDetails = new UploadDetails(username, Instant.now());
+
+            var originalArtifacts = existingPublication.getAssociatedArtifacts();
+            var updatedArtifacts = publicationUpdate.getAssociatedArtifacts();
+
+            for (var updatedArtifact : updatedArtifacts) {
+                if (!originalArtifacts.contains(updatedArtifact)) {
+                    if (updatedArtifact instanceof File file) {
+                        var index = updatedArtifacts.indexOf(updatedArtifact);
+                        var fileBuilder = file.copy().withUploadDetails(uploadDetails);
+
+                        if (file instanceof PublishedFile) {
+                            updatedArtifacts.set(index, fileBuilder.buildPublishedFile());
+                        }
+
+                        if (file instanceof UnpublishedFile) {
+                            updatedArtifacts.set(index, fileBuilder.buildUnpublishedFile());
+                        }
+
+                        if (file instanceof AdministrativeAgreement) {
+                            updatedArtifacts.set(index, fileBuilder.buildUnpublishableFile());
+                        }
+                    }
+                }
+            }
+        }
+    }
+
```

Comment:
Nice, can it be refactored to something like this:

```
            var updatedArtifacts = publicationUpdate.getAssociatedArtifacts().stream()
                                       .filter(updatedArtifact -> !originalArtifacts.contains(updatedArtifact))
                                       .filter(File.class::isInstance)
                                       .map(File.class::cast)
                                       .map(file -> file.copy().withUploadDetails(uploadDetails).build())
                                       .collect(Collectors.toList());
```
        
I see that we do not have build() method in `File.class`, but it would be convinient to implement one:

```
public File buildFile(Class<? extends File> clazz) {
    if (clazz.equals(PublishedFile.class)) {
        return new PublishedFile(identifier, name, mimeType, size, license, administrativeAgreement,
                                publisherVersion, embargoDate, rightsRetentionStrategy,
                                legalNote, Instant.now(), uploadDetails);
    } else if (clazz.equals(UnpublishedFile.class)) {
        return new UnpublishedFile(identifier, name, mimeType, size, license, administrativeAgreement,
                                  publisherVersion, embargoDate, rightsRetentionStrategy, legalNote, uploadDetails);
    } else if (clazz.equals(AdministrativeAgreement.class)) {
        return new AdministrativeAgreement(identifier, name, mimeType, size, license, administrativeAgreement,
                                          publisherVersion, embargoDate, uploadDetails);
    } else {
        throw new IllegalArgumentException("Invalid file class: " + clazz.getName());
    }
}
```

So you could than do something like:

`.map(file -> file.copy().withUploadDetails(uploadDetails).build(file.getClass()))`